src/air/constraints/boundary.rs

```diff
@@ -67,17 +64,29 @@ impl<F: IsField> BoundaryConstraints<F> {
         primitive_root: &FieldElement<F>,
         count_cols_trace: usize,
     ) -> Vec<Vec<FieldElement<F>>> {
-        let mut ret = Vec::new();
-
-        for i in 0..count_cols_trace {
-            ret.push(
-                self.steps(i)
-                    .into_iter()
-                    .map(|s| primitive_root.pow(s))
-                    .collect(),
-            );
-        }
-        ret
+        let max_s: usize = (0..count_cols_trace)
+            .map(|i| {
+                self.constraints
+                    .iter()
+                    .filter(move |v| v.col == i)
+                    .map(|c| c.step)
+            })
+            .flatten()
+            .max()
+            .unwrap_or_default();
+        let roots: Vec<_> =
+            core::iter::successors(Some(FieldElement::one()), |x| Some(x * primitive_root))
+                .take(max_s + 1)
+                .collect();
```

Comment:
Is this really faster? This will always be computing all powers of `primitive_root`. Usually only a handful of those are needed. Maybe I'm getting it wrong, but I think the square and multiply algorithm that uses `pow` in the previous implementation should be faster than computing all of them in order.

Reply:
The goal here is recycling the results. Paraphrasing our chat the other day (correct me if I say anything wrong), a possibly more efficient solution would be to get a sorted, unique list of the actual powers we need, and then accumulate exactly that amount by computing `x^(e[j]-e[i])` where `x` is the primitive root, `e` is the array of exponents we're interested in, and `i` and `j` the respective indices on that array.
Is this correct?

Reply:
Then, all the powers will be the product of consecutive results in the output vector.

Reply:
yes, exactly!